### PR TITLE
Intoduce JdbcDogDAO

### PR DESCRIPTION
**Pull request by gorshkovayulia**
Intoduce JdbcDogDAO, add Add H2 JDBC Driver, implement all methods with Statement class, improve tests

### CODE REVIEW

**Comment by gorshkovayulia** [comment]
@ctapobep please review

File: pom.xml [hunk 1]
```diff
@@ -18,6 +18,12 @@
             <artifactId>spring-webmvc</artifactId>
             <version>4.3.30.RELEASE</version>
         </dependency>
+        <dependency>
+            <groupId>org.springframework</groupId>
+            <artifactId>spring-jdbc</artifactId>
+            <version>5.3.9</version>
+            <scope>compile</scope>
```
Comment (ctapobep) [inline pom.xml]: This is the default scope, we usually don't specify it explicitly.

File: src/main/java/com/example/dog/dao/InMemoryDogDAO.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    @Override
+    public Dog add(Dog dog) {
+        id.incrementAndGet();
+        dog.setId(id.get());
+        dogs.put(dog.getId(), dog);
+        return dog;
+    }
+
+    @Override
+    public Dog update(int id, Dog dog) {
```
Comment (ctapobep) [inline src/main/java/com/example/dog/dao/InMemoryDogDAO.java:27]: This logic looks awkward

File: src/main/java/com/example/dog/dao/JdbcDogDAO.java [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+
+    @Override
+    public Dog get(int id) {
+        Statement statement;
+        ResultSet resultSet;
+        Dog returnedDog = null;
+        try (Connection conn = dataSource.getConnection()) {
+            statement = conn.createStatement();
+            resultSet = statement.executeQuery("select id, name, height, weight, birthday from dog where id = " + id);
+            while (resultSet.next()) {
```
Comment (ctapobep) [inline src/main/java/com/example/dog/dao/JdbcDogDAO.java]: `while` or `if`? If there's more than 1 dog returned, I'd be surprised and would rather expect an error.

File: src/main/java/com/example/dog/dao/JdbcDogDAO.java [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+    }
+
+    @Override
+    public Dog add(Dog dog) {
+        Statement statement;
+        try (Connection conn = dataSource.getConnection()) {
+            statement = conn.createStatement();
+            ZonedDateTime date = dog.getDateOfBirth();
+            String formattedDate = date.format(DateTimeFormatter.ofPattern("uuuu-MM-dd HH:mm:ss VV"));
+            String dateString = formattedDate == null ? null : "'" + formattedDate + "'";
```
Comment (ctapobep) [inline src/main/java/com/example/dog/dao/JdbcDogDAO.java]: See the warning by InelliJ

File: src/main/java/com/example/dog/dao/JdbcDogDAO.java [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+
+            int affectedRows = statement.executeUpdate("insert into dog (name, height, weight, birthday) values ("
+                    + "'" + dog.getName() + "', " + dog.getHeight() + ", " + dog.getWeight() + ", " + dateString + ")",
+                    Statement.RETURN_GENERATED_KEYS);
+            if (affectedRows == 0) {
+                throw new SQLException("Creating dog failed.");
```
Comment (ctapobep) [inline src/main/java/com/example/dog/dao/JdbcDogDAO.java]: You should probably never throw SQLException yourself. Moreover, you'll catch this exception and then re-throw it couple of lines below.

File: src/main/java/com/example/dog/dao/JdbcDogDAO.java [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+    }
+
+    @Override
+    public Dog update(int id, Dog dog) {
+        Statement statement;
+        Dog updatedDog;
+        try (Connection conn = dataSource.getConnection()) {
+            statement = conn.createStatement();
+            ZonedDateTime date = dog.getDateOfBirth();
+            String formattedDate = date.format(DateTimeFormatter.ofPattern("uuuu-MM-dd HH:mm:ss VV"));
```
Comment (ctapobep) [inline src/main/java/com/example/dog/dao/JdbcDogDAO.java]: Looks like duplication

File: src/main/java/com/example/dog/dao/InMemoryDogDAO.java [hunk 7]
```diff
@@ -28,9 +28,8 @@ public Dog update(int id, Dog dog) {
         Dog newValue = dogs.computeIfPresent(id, (k, v)-> dog);
```
Comment (ctapobep) [inline src/main/java/com/example/dog/dao/InMemoryDogDAO.java]: I think this line was more awkward. This would look simpler:

```
    public Dog update(int id, Dog dog) {
        if (!dogs.containsKey(id))
            return null;
        dog.setId(id);
        dogs.put(id, dog);
        return dog;
    }
```

File: src/main/java/com/example/dog/dao/JdbcDogDAO.java [hunk 8]
```diff
@@ -124,4 +111,8 @@ public Dog remove(int id) {
     private static ZonedDateTime getDateTime(Timestamp timestamp) {
         return timestamp != null ? ZonedDateTime.ofInstant(timestamp.toInstant(), ZoneId.systemDefault()) : null;
     }
+
+    private String getDateString(ZonedDateTime time) {
+        return time == null ? null : "'" + time.format(DateTimeFormatter.ofPattern("uuuu-MM-dd HH:mm:ss VV")) + "'";
```
Comment (ctapobep) [inline src/main/java/com/example/dog/dao/JdbcDogDAO.java]: You can make a constant out of `DateTimeFormatter.ofPattern("uuuu-MM-dd HH:mm:ss VV")`*, so that you don't have to parse the pattern on each method invocation.

*this new `DateTimeFormatter` is thread-safe as opposed to the old one that worked with `Date`, so it's safe to make a constant out of it

File: src/test/java/com/example/dog/controller/DogControllerRestAssuredMockMvcTest.java [hunk 9]
```diff
@@ -31,10 +31,7 @@ public void setUp() {
 
     @Test
     public void possibleToGetExistingDog() {
-        Dog dog = new Dog("Tuzik", 24, 8,
```
Comment (ctapobep) [inline src/test/java/com/example/dog/controller/DogControllerRestAssuredMockMvcTest.java:24]: Why did you remove the date? You need to check it too..

File: src/main/java/com/example/dog/dao/InMemoryDogDAO.java (hunk 7 above)
Comment (gorshkovayulia) [reply src/main/java/com/example/dog/dao/InMemoryDogDAO.java]: That's interesting...it was your idea to use `computeIfPresent(id, (k, v)-> dog)` method and I don't understand why this line now looks awkward

File: src/test/java/com/example/dog/controller/DogControllerRestAssuredMockMvcTest.java (hunk 9 above)
Comment (gorshkovayulia) [reply src/test/java/com/example/dog/controller/DogControllerRestAssuredMockMvcTest.java:24]: Because, first of all, I got `NullPointerException` If I passed `null` value. I should've checked this case since it was a problem to _format_ such ZonedDateTime when you `post` a dog (now it fixed in code). Secondly, other test cases have a _normal_ date in post requests and I see no reason to duplicate this test with a dog that has a _normal_ date

File: src/main/java/com/example/dog/dao/InMemoryDogDAO.java (hunk 7 above)
Comment (ctapobep) [reply src/main/java/com/example/dog/dao/InMemoryDogDAO.java]: Just more complicated than it should be

File: src/main/resources/controllers-context.xml [hunk 10]
```diff
@@ -6,8 +6,24 @@
...
+
+    <bean id="inMemoryDogDao" class="com.example.dog.dao.InMemoryDogDAO">
+    </bean>
+
+    <bean id="jdbcDogDao" class="com.example.dog.dao.JdbcDogDAO" primary="true">
+        <constructor-arg ref="dataSource"/>
+    </bean>
+
+    <bean id = "dataSource" class = "org.springframework.jdbc.datasource.DriverManagerDataSource">
+        <property name = "driverClassName" value = "org.h2.Driver"/>
```
Comment (ctapobep) [inline src/main/resources/controllers-context.xml:21]: After you push to master, let's make these configurable properties. Move the values into some kind of `application.properties` and use placeholders here.

**Comment by ctapobep** [comment]
+2
